iothub.class.nut

```diff
@@ -5,724 +5,1024 @@
+            return "?api-version=2016-11-14";
+        }
     }
 
+    //------------------------------------------------------------------------------
+
+    Device = class {
 
-    function toString() {
-        // The toString method shall return a shared-access signature token of the form:
-        // SharedAccessSignature sr=<resourceUri>&sig=<urlEncodedSignature>&se=<expiry>&skn=<urlEncodedKeyName>
-        local sas = "SharedAccessSignature ";
-        foreach (key in ["sr", "sig", "skn", "se"]) {
-            // The skn segment is not part of the returned string if the skn property is not defined.
-            if (this[key]) {
-                if (sas[sas.len() - 1] != ' ') sas += "&";
-                sas += key + "=" + this[key];
+        _body = null;
+
+        constructor(jsonData = null) {
+
+            if (jsonData) {
+                _body = http.jsondecode(jsonData);
+            } else {
+                _body = {
+                    "deviceId": split(http.agenturl(), "/").pop(),
+                    "generationId": null,
+                    "etag": null,
+                    "connectionState": "Disconnected",
+                    "status": "Enabled",
+                    "statusReason": null,
+                    "connectionStateUpdatedTime": null,
+                    "statusUpdatedTime": null,
+                    "lastActivityTime": null,
+                    "cloudToDeviceMessageCount": 0,
+                    "authentication": {
+                        "symmetricKey": {
+                            "primaryKey": null,
+                            "secondaryKey": null
+                        }
+                    }
+                };
             }
         }
 
-        return sas;
+        function connectionString(hostname) {
+            // NOTE: This method did not appear in the original Node.js SDK
+            return format("HostName=%s;DeviceId=%s;SharedAccessKey=%s", hostname, _body.deviceId, _body.authentication.symmetricKey.primaryKey);
+        }
+
+        function getBody() {
+            return _body;
+        }
+
+        function _typeof() {
+            return "Device";
+        }
     }
-}
 
-//------------------------------------------------------------------------------
-class iothub.HTTP {
+    Message = class {
 
-    _config = null;
+        static VALUE = "string"; // message type for auth requests
+        static DATA = "binary"; // message type for stream analytics jobs
 
-    static FEEDBACK_ACTION_ABANDON = "abandon";
-    static FEEDBACK_ACTION_REJECT = "reject";
-    static FEEDBACK_ACTION_COMPLETE = "complete";
+        _body = null;
+        // these are application set properties, not the message properties set by azure
+        _properties = null;
 
-    static HTTP_RETRY_TIME = 10.8; // 24*60*60 (seconds in a day) / 8000 (events per day)
+        constructor(body, properties = null) {
+            _body = body;
+            _properties = properties;
```

Comment:
Is it safe to keep it by reference? What if the app changes it from outside?

I'm not sure if it's worth it, but it will probably be safer to clone it.